cypress-io/cypress
# fix: CSS import in CT Support file is not working

- Closes #23025 

### User facing changelog

Webpack >=4 adds a new feature: giving the user the ability to add sideEffects to their package.json files. If a user adds the CT Support file as a part of their side effect, all global application styles that may have been imported in the support file (e.g. import ./example.css) are not applied. 
- The change implemented ensures that all css files are flagged as containing side Effects and thus are not tree-shaken. This will ensure that there will be no dead-code removal of css files by Webpack and so all styles imported to the CT Support file will be enforced in the user's component irrespective of the sideEffects set by the user.

### Additional details

- Why the problem exists: Webpack 4 introduced several optimizations including dead code removal through tree shaking. Webpack follows all your import and export statements keeping track of the libraries and functions you use and don’t use. So, when it sees something is being imported but not being used, it considers that as ‘dead-code’ and will tree-shake it. But dead code is not always clear though. In our case, the CT support file imports *.css files but never utilizes them. Webpack believes this is dead code and therefore removes them. But this is inherently wrong. Some imports, simply by being included in the bundle have an important impact on the application e.g. global stylesheets and global configurations (in our case *.css and *.command.js are examples of this).
- Other considered solutions:
1. Scaffolding the user's package.json file and directly editing the sideEffects flag from there. This could be messy and could lead to a lot of architecture changes with how we interact and consume the package.json
2. Setting sideEffects: true globally in the module.rules. This tells Webpack that all files contain sideEffects preventing any file from being tree shaken increasing bundle size and slowing down the application.
3. Add optimization.sideEffects: false. This means that any sideEffects flag set in the package.json will be ignored. Issue is that if the user needs some files to be protected from tree shaking while running cypress, setting this to false could lead to a situation where the user’s sideEffects flag in their package.json is ignored and therefore their intended use case is not met.
- Implemented solution: Set sideEffects: true for css modules through module.rules. In our case we want all css files to have a flag of sideEffects: true. This solution is better than setting global sideEffects: true as only css files will be prevented from tree shaking as compared to all the files in the latter option.

### Steps to test

1. Change directory to webpack-dev-server
2. Run yarn cypress:open
3. Run the react.cy.ts testing file. All the projects tested within this file have been configured to include sideEffects in their respective package.json. I have included an assertion to ensure that the background color is applied to the component after the global css files are imported into the CT Support file.

### How has the user experience changed?

Before:
<img width="504" alt="Screen Shot 2022-10-04 at 12 22 07 PM" src="https://user-images.githubusercontent.com/65267668/193872931-359ba0c0-8482-4722-9816-7d72e4348418.png">

After:
<img width="507" alt="Screen Shot 2022-10-04 at 12 23 02 PM" src="https://user-images.githubusercontent.com/65267668/193873086-89474d52-7377-4848-8a32-4ba8dde782b9.png">

### PR Tasks


- [na] Has the original issue (or this PR, if no issue exists) been tagged with a release in ZenHub? (user-facing changes only)
- [na] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [na] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?
